node/impl/full/state_test.go

```diff
@@ -0,0 +1,282 @@
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/network"
+
+	"github.com/filecoin-project/lotus/chain/actors/policy"
+	"github.com/filecoin-project/lotus/chain/beacon"
+	"github.com/filecoin-project/lotus/chain/consensus/filcns"
+	"github.com/filecoin-project/lotus/chain/gen"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/node/impl/full"
+)
+
+func init() {
+	policy.SetSupportedProofTypes(abi.RegisteredSealProof_StackedDrg2KiBV1)
+	policy.SetConsensusMinerMinPower(abi.NewStoragePower(2048))
+	policy.SetMinVerifiedDealSize(abi.NewStoragePower(256))
+}
+
+// similar to chain/rand/rand_test.go
+func TestStateGetBeaconEntry(t *testing.T) {
+	// Ref: https://github.com/filecoin-project/lotus/issues/12414#issuecomment-2320034935
+	type expectedBeaconStrategy int
+	const (
+		expectedBeaconStrategy_beforeNulls expectedBeaconStrategy = iota
+		expectedBeaconStrategy_afterNulls
+		expectedBeaconStrategy_exact
+	)
+
+	testCases := []struct {
+		name          string
+		nv            network.Version
+		strategy      expectedBeaconStrategy // how to determine which round to expect
+		wait          bool                   // whether the test should wait for a future round
+		negativeEpoch bool
+	}{
+		{
+			// In v12 and before, if the tipset corresponding to round X is null, we fetch the latest beacon entry BEFORE X that's in a non-null ts
+			name:     "pre-nv12@1 nulls",
+			nv:       network.Version1,
+			strategy: expectedBeaconStrategy_beforeNulls,
+		},
+		{
+			name:     "pre-nv12@9 nulls",
+			nv:       network.Version9,
+			strategy: expectedBeaconStrategy_beforeNulls,
+		},
+		{
+			name:     "pre-nv12@10 nulls",
+			nv:       network.Version10,
+			strategy: expectedBeaconStrategy_beforeNulls,
+		},
+		{
+			name:     "pre-nv12@12 nulls",
+			nv:       network.Version12,
+			strategy: expectedBeaconStrategy_beforeNulls,
+		},
+		{
+			name:     "pre-nv12 wait for future round",
+			nv:       network.Version12,
+			strategy: expectedBeaconStrategy_exact,
+			wait:     true,
+		},
+		{
+			name:          "pre-nv12 requesting negative epoch",
+			nv:            network.Version12,
+			negativeEpoch: true,
+		},
+		{
+			// At v13, if the tipset corresponding to round X is null, we fetch the latest beacon entry in the first non-null ts after X
+			name:     "nv13 nulls",
+			nv:       network.Version13,
+			strategy: expectedBeaconStrategy_afterNulls,
+		},
+		{
+			name:          "nv13 requesting negative epoch",
+			nv:            network.Version13,
+			negativeEpoch: true,
+		},
+		{
+			name:     "nv13 wait for future round",
+			nv:       network.Version13,
+			strategy: expectedBeaconStrategy_exact,
+			wait:     true,
+		},
+		{
+			// After v14, if the tipset corresponding to round X is null, we still fetch the randomness for X (from the next non-null tipset) but can get the exact round
```

Comment:
no, that's the 20 iterations thing that needs cleaning up:

1. call `getBeaconEntryV3` with the filecoin epoch we want
2. call `GetBeaconRandomnessTipset` on that epoch with `lookback` set to `false`, so it'll skip ahead to the next non-null tipset
3. look in the beacon entries in that tipset for the precise drand round number that we want
4. if we don't find it, walk back up the chain ... but we're not going to find it up the chain because we're matching the exact drand round number so it'll always end in an error, just after a 20 tipset walkback